qiskit_nature/second_q/operators/polynomial_tensor.py

```diff
@@ -0,0 +1,119 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Polynomial Tensor class"""
+
+from __future__ import annotations
+from typing import Dict
+from numbers import Number
+import numpy as np
+from qiskit.quantum_info.operators.mixins import (
+    LinearMixin,
+    AdjointMixin,
+    TolerancesMixin,
+)
+
+
+class PolynomialTensor(LinearMixin, AdjointMixin, TolerancesMixin):
+    """Polynomial Tensor class"""
+
+    def __init__(self, data: Dict[str, np.ndarray]):
+        if not isinstance(data, Dict):
+            raise TypeError(f"data {data} must be a Dictionary")
+
+        self._data = data
+
+        shapes = ()
```

Comment:
I think it would be useful, after the following verification, to expose the final `shape` value as an attribute of this class.
This could then be used during e.g. `add` to ensure that the shapes of two instances match, without having to check all of the internal matrices (again)

Reply:
We need to figure what we want to do to expose the data. As far as checking need for operations, add, mul, etc since the instances should each be valid, it only needs one matrix to be checked against one in the other to check size of dimension matches - then they either all match or they all don't, since they all should match inside a valid instance

Reply:
I think having access to the `size` (which later translates to the `register_length` and even later to the `num_qubits`) will be useful also during the `SparseLabelOp.from_polynomial_tensor` implementations.

As to exposing `_data`, I would have defined an `__iter__` method which allows iteration over `_data.items()`.
Since `PolynomialTensor` is supposed to be an immutable object, read-only access by means of iterating the internal data should be sufficient.
Although we may also to expose access to individual objects in `_data`, but that would require the user to know the keys he would need to use for accessing..

Reply:
Or we can expose keys() and let the user access any entry they want from a key. This would expose the matrix directly by reference (as would an iter I think) so immutability will be there in as much as its treated as read-only. Since its constructed with a dict being able to later access part of it by key seems consistent. (I had thought the constant would have its own access if it were separate)